### PR TITLE
MCLOUD-3025: Sitemap generation warnings

### PR DESCRIPTION
**Pull request by o-iegorov**
Fix for preventing sitemap generation warnings

### Fixed Issues (if relevant)
1. https://jira.corp.magento.com/browse/MCLOUD-3025

### Manual testing scenarios
Check that background sitemap generation still working as expected

**Negative test scenario**

Change the fs access right to read only fro the folder with sitemaps. After cron job run check that debug logs contains generation errors and retry attempts

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

releasenotes
Added retry attempt when errors occur during sitemap generation to skip customer email notification in cases where errors can be recovered automatically.

### CODE REVIEW

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
++                try {
++                    $sitemap->generateXml();
++                } catch (\Exception $e) {
++                    $nonRecoverableErrors[] = $e->getMessage();
++                }
++            }
++            $this->emailNotification->sendErrors($nonRecoverableErrors);
+         }
+     }
+ }
```
Comment (NadiyaS) [inline patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: add empty line

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+             try {
+                 $sitemap->generateXml();
+             } catch (\Exception $e) {
++                $sitemapsWithError[] = $sitemap;
+                 $errors[] = $e->getMessage();
+             }
+         }
+-        if ($errors && $recipient) {
+-            $this->emailNotification->sendErrors($errors);
++        if ($errors && $recipient && !empty($sitemapsWithError)) {
```
Comment (NadiyaS) [inline patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Why did you add `!empty($sitemapWithError)` in condition? Is it possible to have empty `$sitemapWithError` and not empty `$errors`?

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
++            $nonRecoverableErrors = [];
++            //re-try to write sitemap again
++            foreach ($sitemapsWithError as $sitemap) {
++                try {
++                    $sitemap->generateXml();
++                } catch (\Exception $e) {
++                    $nonRecoverableErrors[] = $e->getMessage();
++                }
++            }
++            $this->emailNotification->sendErrors($nonRecoverableErrors);
```
Comment (NadiyaS) [inline patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch:45]: I would also add logs about error with sitemap generation and about the second try

**Review by NadiyaS — CHANGES_REQUESTED** [review]
I would also add logs about en error with sitemap generation and about the second try

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch (hunk 2 above)
Comment (o-iegorov) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Redundant code removed

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch (hunk 1 above)
Comment (o-iegorov) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Are you sure that empty line is needed at the patch not at the end of the php file?

**Comment by o-iegorov** [comment]
> I would also add logs about en error with sitemap generation and about the second try

Logging already added

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch (hunk 1 above)
Comment (NadiyaS) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Yes, I am sure. Actually, it should be added by default when you create patch, so it is strange that you do not have it.

File: patches.json [hunk 4]
```diff
@@ -238,6 +238,9 @@
     },
     "Fix PayPal issue with region": {
       "100.3.4": "MC-31387__fix_paypal_issue_with_region__100.3.4.patch"
+    },
+    "Sitemap Generation Warnings": {
+      "2.3.3": "MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch"
```
Comment (NadiyaS) [inline patches.json]: oops, just noticed that you added this to paypal. Please assign this patch to magento2-base

Comment (o-iegorov) [reply patches.json]: Moved to the proper position

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch (hunk 1 above)
Comment (o-iegorov) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: I have it? but it's not reflected in review, as well as in any other reviews for patches:

Comment (o-iegorov) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: https://github.com/magento/magento-cloud-patches/pull/35/files
https://github.com/magento/magento-cloud-patches/pull/34/files
https://github.com/magento/magento-cloud-patches/pull/32/files

Comment (NadiyaS) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Extra empty line at the end is not displayed in the PR, but if you don't have, you will see that red mark as you had before in this file.

File: patches.json [hunk 5]
```diff
@@ -230,6 +230,9 @@
     "FPC is getting disabled during deployments": {
       ">=2.3.2 <2.3.5": "MAGECLOUD-5069__fpc_is_getting_disabled_during_deployments__2.3.2.patch",
       ">=2.3.5 <2.4.0": "MAGECLOUD-5069__fpc_is_getting_disabled_during_deployments__2.3.5.patch"
+    },
+    "Sitemap Generation Warnings": {
+      ">=2.3.3 <2.3.5": "MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch"
```
Comment (NadiyaS) [inline patches.json]: We have 3 customers who reported about this problem, on that moment they had 2.3.1 and 2.3.3 Magento versions. 
As this problem reproducible only on cloud project, customers on any version may have this issue, but I have concerns about increasing version range.
@YPyltiai your suggestions?

**Review by shiftedreality — DISMISSED** [review]

File: patches.json (hunk 5 above)
Comment (YPyltiai) [reply patches.json]: @NadiyaS if we know that this reproduces anywhere else beyond this range - we may need to cover that. We don't want to re-release packages next time it happens on 2.3.1. Can we somehow validate it, please?

Comment (o-iegorov) [reply patches.json]: Added 2.3.1 inclusively to cover all active cases

Comment (NadiyaS) [reply patches.json]: That is the problem that we cannot reproduce it. And even on the client projects it is a random error.
This patch does not fix the problem itself. It is a workaround which we do not want to deliver everywhere.
so, we will deliver starting 2.3.1 then?

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+             } catch (\Exception $e) {
++                $sitemapsWithError[] = $sitemap;
+                 $errors[] = $e->getMessage();
+             }
+         }
++
+         if ($errors && $recipient) {
+-            $this->emailNotification->sendErrors($errors);
++            sleep($this->retryInterval);
++            $this->logger->debug('Sitemap generation retry occurred');
```
Comment (NadiyaS) [inline patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: I would log the error code itself too. and then that we retry generation.
And at the end add error to logs if it is still present (that one which customer will receive)

File: patches.json (hunk 5 above)
Comment (YPyltiai) [reply patches.json]: @NadiyaS I honestly see no harm in re-try operation - it will trigger only when initially it failed, which will not affect anyone who has no issues. So I would cover everyone who is affected. At least from my point of view I don't see what else can it break but maybe I am missing something.

File: patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch (hunk 6 above)
Comment (o-iegorov) [reply patches/MCLOUD-3025__sitemap_generation_warnings__2.3.3.patch]: Improved logging implemented

**Review by NadiyaS — APPROVED** [review]

**Review by shiftedreality — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Comment by andriyShevtsov** [comment]
@o-iegorov Please add to description about email sending in case of error. Looks like it important part of this fix
